Use explicit checks for arrays in boolean contexts

schuetzm opened the pull request:
Newest DMD is more picky about implicit conversion of arrays to bool, see this PR:
https://github.com/D-Programming-Language/dmd/pull/2885

s-ludwig commented on source/vibe/data/bson.d:
```diff
@@ -1428,7 +1428,7 @@ struct BsonSerializer {
 		}
 
-		if (m_entryName) {
+		if (m_entryName != "") {
```
Still not perfect, but this should better be `!is null`, because dictionary keys can in theory be `""`, too.

s-ludwig commented:
Otherwise LGTM.

schuetzm commented:
Added a bunch of others that turned up when I removed a now unnecessary cast (PR coming). There are still a few left that I can't locate:

```
/home/marc/d/phobos/std/exception.d(351): Warning: implicit conversion of dynamic arrays to bool can be ambiguous and will be deprecated. Use one of: value !is null, value.length, or value.ptr instead
__diet_code__(11): Warning: implicit conversion of dynamic arrays to bool can be ambiguous and will be deprecated. Use one of: "foo" !is null, "foo".length, or "foo".ptr instead
__diet_code__(11): Warning: implicit conversion of dynamic arrays to bool can be ambiguous and will be deprecated. Use one of: cast(string)null !is null, cast(string)null.length, or cast(string)null.ptr instead
__diet_code__(11): Warning: implicit conversion of dynamic arrays to bool can be ambiguous and will be deprecated. Use one of: "foo" !is null, "foo".length, or "foo".ptr instead
__diet_code__(11): Warning: implicit conversion of dynamic arrays to bool can be ambiguous and will be deprecated. Use one of: cast(string)null !is null, cast(string)null.length, or cast(string)null.ptr instead
```

schuetzm commented:
Ok, found the remaining ones.

s-ludwig commented on source/vibe/web/validation.d:
```diff
@@ -76,7 +76,7 @@ struct ValidUsername {
 
 		auto err = appender!string(); // TODO: avoid allocations when possible
-		if (validateUserName(str)) ret = ValidUsername(str);
+		if (validateUserName(str) != "") ret = ValidUsername(str);
```
Good that these lines appear now in a diff. They are using the wrong overload. Should be:
`if (validateUserName(err, str)) ret = ValidUsername(str);`

I wouldn't have expected the DMD change to catch these kinds of bugs, too :)

s-ludwig commented on source/vibe/web/validation.d:
```diff
@@ -120,7 +120,7 @@ struct ValidPassword {
 
 		auto err = appender!string(); // TODO: avoid allocations when possible
-		if (validatePassword(str, str)) ret = ValidPassword(str);
+		if (validatePassword(str, str) != "") ret = ValidPassword(str);
```
Same here.

s-ludwig commented:
The `getInterfaceValidationError` comparisons should probably use `!is null` instead to be consistent with the rest of the module. Although in reality both should always be equivalent. @Geod24, any preference?

schuetzm commented:
Changed the validation functions to use the output range overloads.

I replaced the `!= ""` by `!is null`. For public functions, I would prefer the string comparison, but as `getInterfaceValidationError` is private, we're in control of all uses and can make sure that it actually never returns `""`.

s-ludwig commented:
Thanks! BTW, I agree in principle with avoiding `!is null`, it would be better to use `Nullable!string` in cases where a "null" is expected. I suspect that `!= ""` is less efficient than `.length` or `!str.empty`, though (has anyone actually measured that by chance?), so I'd usually rather use the latter (but no need to do that for the changes here).

schuetzm commented:
It still compiles to way too much code:
https://issues.dlang.org/show_bug.cgi?id=14436

Before yebblies fixed that, testing for `null` was just as bad.

But this is clearly something that should be fixed in the compiler.
